### PR TITLE
test[CL]: CurrentSqrtPrice

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #3239 

## What is the purpose of the change

This PR introduces a check in both Calc/Swap methods (for both In and Out) that the currentSqrtPrice we produce is what we expect. 

These checks took me down a bit of a rabbit hole to finally discover that the OOO was wrong on one of our liquidity checks, which is why the expected liquidity for a lot of the test cases has now changed. 

## Brief Changelog

  - Utilizes osmomath in Liquidity0, Liquidity1, and TickToSqrtPrice methods
  - Modfies Calc/Swap tests for both In and Out to utilize newer, more precise liquidity values as well as check that our final sqrtPrice is as intended (with corresponding wolframalpha links

### CODE REVIEW

File: x/concentrated-liquidity/internal/math/math.go
Comment (mattverse) [inline x/concentrated-liquidity/internal/math/math.go:112]: Confirmed that there is no logic change here

File: x/concentrated-liquidity/internal/math/math.go
Comment (mattverse) [inline x/concentrated-liquidity/internal/math/math.go:104]: Wdyt about creating an issue for this regarding adding testcase for testing product overflow?

File: x/concentrated-liquidity/internal/math/math_test.go
Comment (mattverse) [inline x/concentrated-liquidity/internal/math/math_test.go:161]: wow im surprised that we're getting so close to expected!

**Review by mattverse — APPROVED** [review]
LGTM. Overall very surprised that we got so close to reducing margin of error just by changing the Dec part for currentSqrtPrice! This is exciting!

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/math.go:18]: Is precision loss significant enough to require this?

I'm surprised that basic math operations are causing a large precision loss

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/math.go:20]: We should do the conversions from `Dec` to `BigDec`

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/math.go:25]: We should not construct this from a string. There should be helpers to perform the conversion `Int` -> `Dec` -> `BigDec`

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/math.go:39]: please change conversions

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/math.go:43]: please change conversions

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/math.go:112]:
```suggestion
	return liquidity.QuoRoundUp(liquidity.Quo(sqrtPriceCurrent).Add(amountRemaining))
```

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/math.go:103]: What is the precise formula?

Why would product overflow?

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/math.go:111]: What do you mean by overflow? I think the definition is causing my confusion here

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/math.go:98]: Can we remove pseudocode in the spec please and, instead, explain what that means?

File: x/concentrated-liquidity/internal/math/tick.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/tick.go:33]: Let's initialize `osmomath.MustNewDecFromStr("1.0001")` as a package-global variable please

File: x/concentrated-liquidity/internal/math/tick.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/tick.go:35]: and reuse it ere

File: x/concentrated-liquidity/internal/math/tick_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/tick_test.go:16]: The values seem slightly off judging by the old comment.

I think we should specify `sqrtPriceExpected` as given by Wolfram and define an "additive tolerance" for each test case.

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [reply x/concentrated-liquidity/internal/math/math.go:18]: I think using `osmomath.BigDec` is useful in `TickToSqrtPrice` since it uses functions such as `Power`. However, I would be surprised if multiplications, subtractions and additions would be the cause of precision loss here

Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:18]: Screenshot 2022-12-19 at 10 34 01 AM

Reverting this back to what it was produces very minimal significance, I was just trying to match the exact result of wolframalpha. Do you think I should revert this to what is was (no using osmomath) and just accept the extremely small precision loss?

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:43]: Offline, agreed to revert liquidity0 and liquidity1 back to normal sdk.Decs due to minimal improvement in precision

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:39]: Offline, agreed to revert liquidity0 and liquidity1 back to normal sdk.Decs due to minimal improvement in precision

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:25]: Offline, agreed to revert liquidity0 and liquidity1 back to normal sdk.Decs due to minimal improvement in precision

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:20]: Offline, agreed to revert liquidity0 and liquidity1 back to normal sdk.Decs due to minimal improvement in precision

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:18]: Offline, agreed to revert liquidity0 and liquidity1 back to normal sdk.Decs due to minimal improvement in precision

File: x/concentrated-liquidity/internal/math/tick_test.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/tick_test.go:16]: old comment was actually off (rounded instead of truncated like we want).

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:18]: After further tests, the precision loss is actually significant enough to warrant utilize BigDecs here. I will add them back and we can discuss further

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:95]: Having the formulas in the godoc has been really helpful for me to reference. I have referenced quite a few times.

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:107]: This logic already existed prior to this PR. I was just adding comments from the uniswap code. The todo here is seeing if this less accurate formula to prevent overflows is needed in our context

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:100]: Precise formula is the top one, bottom one is less precise.

I don't know solidity math well enough to answer why their product might overflow

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go]: https://github.com/osmosis-labs/osmosis/issues/3791

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [inline x/concentrated-liquidity/internal/math/math.go:17]: Let's add a comment why we are making these conversions to `BigDec` please. That is, to avoid frequent precision loses for callers of this function

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [reply x/concentrated-liquidity/internal/math/math.go:101]: It doesn't seem like a formula, more like a code block/pseudocode which shouldn't be in the spec.

The spec should abstract away the implementation details, allowing the developer to treat the function as a black box. In the current state, it only feels like additional complexity trying to visually QA the spec first before reading the function implementation. 

I agree that formulas are useful. However, if we are to leave them in the codebase, it should be clear what they are referring to exactly. I'm supportive of having them if they are clearly referencing the code lines and we don't have to include large chunks of pseudocode in the spec

File: x/concentrated-liquidity/internal/math/math.go
Comment (p0mvn) [reply x/concentrated-liquidity/internal/math/math.go:106]: Let's not block this PR but create an issue to re-think this entire function and understand the math / whether we need the two kinds of computations

**Review by p0mvn — APPROVED** [review]
Pre-approving for when comments for `GetNextSqrtPriceFromAmount0RoundingUp` are addressed.

In this PR, I think it would be valuable to clean up the comments for that function.

In a future PR / issue, it would be valuable to revisit the calculations for that function, understand why we have two kinds and potentially simplify it

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:113]: https://github.com/osmosis-labs/osmosis/issues/3887

File: x/concentrated-liquidity/internal/math/math.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/internal/math/math.go:97]: Removed

**Pull request merged**
